cwr/index.go

```diff
@@ -71,55 +87,259 @@ func (i *Indexer) Index(ctx context.Context, stream chan *cid.Cid) error {
 	}
 }
 
-// index indexes a cwr records based on the record type (NWR/REV/SPU)
-func (i *Indexer) index(cwrRecord *meta.Object) error {
+// index indexes a CWR based on its NWR,REV
+func (i *Indexer) index(cwr *meta.Object) (err error) {
+	graph := meta.NewGraph(i.store, cwr)
+	jobs := make(chan jobIn)
+	results := make(chan error)
+	wg := new(sync.WaitGroup)
 
-	recordType, err := cwrRecord.GetString("record_type")
-	if err != nil {
-		return err
-	}
-	if recordType == "NWR" || recordType == "REV" {
-
-		registeredWork := &RegisteredWork{}
-		if err := cwrRecord.Decode(registeredWork); err != nil {
+	for field, indexFn := range map[string]func(*cid.Cid, *meta.Object) error{
+		"HDR": i.indexTransmissionHeader,
+		"TRL": i.indexTransmissionTrailer,
+	} {
+		v, err := graph.Get("Records", field)
+		if meta.IsPathNotFound(err) {
+			continue
+		} else if err != nil {
 			return err
 		}
-		if err := i.indexRegisteredWork(cwrRecord.Cid().String(), registeredWork); err != nil {
+		id, ok := v.(*cid.Cid)
+		if !ok {
+			return fmt.Errorf("unexpected field type for %q, expected *cid.Cid, got %T", field, v)
+		}
+		if err := i.indexRecord(cwr.Cid(), id, indexFn); err != nil {
 			return err
 		}
 	}
-	if recordType == "SPU" {
+	v, err := graph.Get("Groups")
+	if err != nil {
+		return err
+	}
 
-		publisherControlledBySubmitter := &PublisherControllBySubmitter{}
-		if err := cwrRecord.Decode(publisherControlledBySubmitter); err != nil {
-			return err
+	numberOfGroups := len(v.([]interface{}))
+
+	for w := 1; w <= concurrentWorkNum; w++ {
+		wg.Add(1)
+		go i.worker(jobs, results, wg)
```

Comment:
As [mentioned previously](https://github.com/meta-network/go-meta/pull/24#discussion_r142787018), let's avoid passing `WaitGroups` around and just have all the `WaitGroup` related code in the same place.